### ISSUE TITLE
setDefaultNamespace() bug

### ISSUE THREAD

**Issue by KarelWintersky** [skipperbent/simple-php-router#446]
Default namespace DOES not applied to all routes if declared after them.

```
SimpleRouter::group(['prefix' => '/:auth'], function (){
    SimpleRouter::get('/', 'Auth@form_login');
    SimpleRouter::get('/login', 'Auth@form_login');
    SimpleRouter::post('/login', 'Auth@callback_login');
    SimpleRouter::get('/logout', 'Auth@callback_logout');
});

SimpleRouter::group([
    'prefix' => '/:admin', 
    'middleware' => \FactGroup\Middleware\CheckAuth::class
], function (){
    SimpleRouter::get('/', 'AdminPages@page_frontpage');
    SimpleRouter::get('/page', 'AdminPages@pages_list');
    
    SimpleRouter::get('/page/add', 'AdminPages@form_add_page');
    SimpleRouter::get('/page/insert', 'AdminActions@callback_page_insert');
    
    SimpleRouter::get('/page/edit/{slug}', 'AdminPages@form_edit_page');
    SimpleRouter::get('/page/update/{slug}', 'AdminActions@callback_page_update');
});

SimpleRouter::get('/', 'Page@page_front');
SimpleRouter::get('/{slug}', 'Page@page_view');

SimpleRouter::setDefaultNamespace('FactGroup\Units');
```
In this case namespace `FactGroup\Units` applied to routes in groups with prefixes `:admin` and `:auth`, but NOT applied to 'simple' routing rules outide group. 

Composer trying search for class `Page` outside `FactGroup\Units` namespace and can't find something. 
It can be solved with declaring full path to route callback:

```
SimpleRouter::get('/', 'FactGroup\Units\Page@page_front');
SimpleRouter::get('/{slug}', 'FactGroup\Units\Page@page_view');
```

but is is a bad solution.

**Comment by KarelWintersky** [comment]
`SimpleRouter::setDefaultNamespace('FactGroup\Units');`

applies namespace to all non-grouped rules declared AFTER, not ALL rules without any exceptions.

Please, fix.

**Comment by skipperbent** [comment]
Hi.

Thanks for pointing that out - it sounds like a bug.

I'll do some digging and see what I can do. I will keep you posted.

/ Simon

**Comment by KarelWintersky** [comment]
Found a lifehack:

```
SimpleRouter::group([], function (){
   SimpleRouter::get('/', 'Page@page_front');
   SimpleRouter::get('/{slug}', 'Page@page_view');
});
```

**Comment by KarelWintersky** [comment]
bump!

**Comment by michaelvcolianna** [comment]
Sorry to bump but this hasn't been addressed in the documentation.

Documentation example:

```<?php
use Pecee\SimpleRouter\SimpleRouter;

/* Load external routes file */
require_once 'routes.php';

/**
 * The default namespace for route-callbacks, so we don't have to specify it each time.
 * Can be overwritten by using the namespace config option on your routes.
 */

SimpleRouter::setDefaultNamespace('\Demo\Controllers');

// Start the routing
SimpleRouter::start();
```

But this will throw a `Pecee\SimpleRouter\Exceptions\NotFoundHttpException` on any definitions in routes.php with controller actions – ClassLoader.php line 31. (Note, I haven't tried extending but since the example in the documentation goes routes.php, setDefaultNamespace, start I'm guessing it also throws the exception.)

It seems the `SimpleRouter::setDefaultNamespace('\Demo\Controllers);` line should either come before routes.php is included OR at the top of routes.php, before any routes are defined.

In lieu of moving the setDefaultNamespace call, the hack by @KarelWintersky also works. I haven't dug into the code but I'm guessing it's only looking to apply the default namespace to groups, so any 'naked' routes get skipped and hence throw the error.

**Comment by skipperbent** [comment]
> Sorry to bump but this hasn't been addressed in the documentation.
> 
> Documentation example:
> 
> ```
> use Pecee\SimpleRouter\SimpleRouter;
> 
> /* Load external routes file */
> require_once 'routes.php';
> 
> /**
>  * The default namespace for route-callbacks, so we don't have to specify it each time.
>  * Can be overwritten by using the namespace config option on your routes.
>  */
> 
> SimpleRouter::setDefaultNamespace('\Demo\Controllers');
> 
> // Start the routing
> SimpleRouter::start();
> ```
> 
> But this will throw a `Pecee\SimpleRouter\Exceptions\NotFoundHttpException` on any definitions in routes.php with controller actions – ClassLoader.php line 31. (Note, I haven't tried extending but since the example in the documentation goes routes.php, setDefaultNamespace, start I'm guessing it also throws the exception.)
> 
> It seems the `SimpleRouter::setDefaultNamespace('\Demo\Controllers);` line should either come before routes.php is included OR at the top of routes.php, before any routes are defined.
> 
> In lieu of moving the setDefaultNamespace call, the hack by @KarelWintersky also works. I haven't dug into the code but I'm guessing it's only looking to apply the default namespace to groups, so any 'naked' routes get skipped and hence throw the error.

This is correct. The `setDefaultNamespace` is handled by the `SimpleRouter` class and aren't compiled at runtime.

I've changed the behavior so default-namespace are now done when calling `SimpleRouter::start` which should resolve the issue.

**Comment by skipperbent** [comment]
This issue has been resolved in version 4.3.0.0.
Feel free to open another issue if you are still experiencing issues.
